internal/api/provider/oidc.go

```diff
@@ -0,0 +1,163 @@
+			SkipClientIDCheck: true,
+		}
+	}
+
+	verifier := provider.Verifier(config)
+
+	token, err := verifier.Verify(ctx, idToken)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var data *UserProvidedData
+
+	switch token.Issuer {
+	case IssuerGoogle:
+		token, data, err = parseGoogleIDToken(token)
+
+	case IssuerApple:
+		token, data, err = parseAppleIDToken(token)
+
+	default:
+		token, data, err = parseGenericIDToken(token)
+	}
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if token.AccessTokenHash != "" {
+		if err := token.VerifyAccessToken(accessToken); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return token, data, nil
+}
+
+func parseGoogleIDToken(token *oidc.IDToken) (*oidc.IDToken, *UserProvidedData, error) {
+	var claims googleUser
+	if err := token.Claims(&claims); err != nil {
+		return nil, nil, err
+	}
+
+	var data UserProvidedData
+
+	if claims.Email != "" {
+		data.Emails = append(data.Emails, Email{
+			Email:    claims.Email,
+			Verified: claims.IsEmailVerified(),
+			Primary:  true,
+		})
+	}
+
+	if len(data.Emails) <= 0 {
+		return nil, nil, errors.New("provider: Google ID token must contain an email address")
+	}
+
+	data.Metadata = &Claims{
+		Issuer:        claims.Issuer,
+		Subject:       claims.Subject,
+		Name:          claims.Name,
+		Picture:       claims.AvatarURL,
+		Email:         claims.Email,
+		EmailVerified: claims.IsEmailVerified(),
+
+		// To be deprecated
+		AvatarURL:  claims.AvatarURL,
+		FullName:   claims.Name,
+		ProviderId: claims.Subject,
+	}
+
+	if claims.HostedDomain != "" {
+		data.Metadata.CustomClaims = map[string]any{
+			"hd": claims.HostedDomain,
+		}
+	}
+
+	return token, &data, nil
+}
+
+type AppleIDTokenClaims struct {
+	jwt.StandardClaims
+
+	Email string `json:"email"`
+
+	AuthTime       *string `json:"auth_time"`
+	IsPrivateEmail *bool   `json:"is_private_email,string"`
+}
+
+func parseAppleIDToken(token *oidc.IDToken) (*oidc.IDToken, *UserProvidedData, error) {
+	var claims AppleIDTokenClaims
+	if err := token.Claims(&claims); err != nil {
+		return nil, nil, err
+	}
+
+	var data UserProvidedData
+
+	data.Emails = append(data.Emails, Email{
+		Email:    claims.Email,
+		Verified: true,
+		Primary:  true,
+	})
+
+	data.Metadata = &Claims{
+		Issuer:        token.Issuer,
+		Subject:       token.Subject,
+		Email:         claims.Email,
+		EmailVerified: true,
+		ProviderId:    token.Subject,
+		CustomClaims:  make(map[string]any),
+	}
+
+	if claims.IsPrivateEmail != nil {
+		data.Metadata.CustomClaims["is_private_email"] = *claims.IsPrivateEmail
+	}
+
+	if claims.AuthTime != nil {
+		data.Metadata.CustomClaims["auth_time"] = *claims.AuthTime
+	}
+
+	if len(data.Metadata.CustomClaims) < 1 {
+		data.Metadata.CustomClaims = nil
+	}
+
+	return token, &data, nil
+}
+
+func parseGenericIDToken(token *oidc.IDToken) (*oidc.IDToken, *UserProvidedData, error) {
+	var data UserProvidedData
+
+	if err := token.Claims(&data.Metadata); err != nil {
```

Comment:
Initialize `data.Metadata` first.

```suggestion
	var data UserProvidedData
	data.Metadata = &Claims{}

	if err := token.Claims(data.Metadata); err != nil {
```